pkg/syncer/admissionhandler/validatepvc.go

```diff
@@ -0,0 +1,198 @@
+				log.Warnf("error deserializing old pvc: %v. skipping validation.", err)
+				return &admissionv1.AdmissionResponse{
+					// skip validation if there is pvc deserialization error
+					Allowed: true,
+				}
+			}
+			newReq = newPVC.Spec.Resources.Requests[corev1.ResourceStorage]
+		} else {
+			reclaimPolicy, err := getReclaimPolicyForPVC(ctx, oldPVC)
+			if err != nil {
+				log.Warnf("error getting reclaim policy for pvc: %v. skipping validation.", err)
+				return &admissionv1.AdmissionResponse{
+					// skip validation if there is any error in getting reclaim policy for pvc
+					Allowed: true,
+				}
+			}
+
+			if reclaimPolicy != corev1.PersistentVolumeReclaimDelete {
+				log.Info("the reclaim policy of PVC is not Delete. skipping validation.")
+				return &admissionv1.AdmissionResponse{
+					// skip validation if the reclaim policy of PVC is not Delete
+					Allowed: true,
+				}
+			}
+		}
+
+		if req.Operation == admissionv1.Delete || req.Operation == admissionv1.Update && newReq.Cmp(oldReq) > 0 {
+			snapshots, err := getSnapshotsForPVC(ctx, oldPVC.Namespace, oldPVC.Name)
+			if err != nil {
+				log.Warnf("error getting snapshots for pvc: %v. skipping validation.", err)
+				return &admissionv1.AdmissionResponse{
+					// skip validation if there is any error in getting volume snapshots associated with the pvc
+					Allowed: true,
+				}
+			}
+			if len(snapshots) != 0 {
+				allowed = false
+				if req.Operation == admissionv1.Update {
+					result = &metav1.Status{
+						Reason: ExpandVolumeWithSnapshotErrorMessage,
+					}
+				} else if req.Operation == admissionv1.Delete {
+					result = &metav1.Status{
+						Reason: DeleteVolumeWithSnapshotErrorMessage,
+					}
+				}
+			}
+		}
+	default:
+		allowed = false
+		log.Errorf("Can't validate resource kind: %q using validatePVC function", req.Kind.Kind)
+	}
+
+	// return AdmissionResponse result
+	return &admissionv1.AdmissionResponse{
+		Allowed: allowed,
+		Result:  result,
+	}
+}
+
+func getReclaimPolicyForPVC(ctx context.Context, pvc corev1.PersistentVolumeClaim) (
+	corev1.PersistentVolumeReclaimPolicy, error) {
+	log := logger.GetLogger(ctx)
+
+	var result corev1.PersistentVolumeReclaimPolicy
+
+	if pvc.Spec.VolumeName == "" {
+		return result, logger.LogNewErrorf(log, "No PV is bound to the PVC %s/%s", pvc.Namespace, pvc.Name)
+	}
+
+	kubeClient, err := k8s.NewClient(ctx)
+	if err != nil {
+		return result, logger.LogNewErrorf(log, "failed to get kube client with error: %v. "+
+			"Stopping getting reclaim policy for PVC, %s/%s", err, pvc.Namespace, pvc.Name)
+	}
+
+	pv, err := kubeClient.CoreV1().PersistentVolumes().Get(ctx, pvc.Spec.VolumeName, metav1.GetOptions{})
+	if err != nil {
+		return result, logger.LogNewErrorf(log, "failed to get PV %v with error: %v. "+
+			"Stopping getting reclaim policy for PVC, %s/%s", err, pvc.Spec.VolumeName, pvc.Namespace, pvc.Name)
+	}
+
+	return pv.Spec.PersistentVolumeReclaimPolicy, nil
```

Comment:
That is a little bit strange here. Actually, I thought about it. But I have to do that. 
Basically, PVC doesn't have reclaimPolicy field with it. So, we have to get the corresponding reclaimPolicy from its bounded PV. If I rename it to getReclaimPolicyForPV, the calling site will be,

```
reclaimPolicy, err := getReclaimPolicyForPV(ctx, oldPVC)
```

That means we are getting some PV property with an input of PVC object, which is also weird. Maybe I should split the current `getReclaimPolicyForPVC` func into `getPVForPVC` and `getReclaimPolicyForPV`.  It might be more reasonable.

Reply:
sure, that would make it much better.

Reply:
I didn't split the func into two. Instead, I just renamed it to `getPVReclaimPolicyForPVC`.